### PR TITLE
[AIRFLOW-2562]Add Google Kubernetes Engine Operators

### PR DESCRIPTION
**Pull request by Noremac201**
### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2562

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

Add Google Kubernetes Engine create_cluster, delete_cluster operators
This allows users to use airflow to create or delete clusters in the
google cloud platform via an operator

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
tests.contrib.operators.test_gcp_container_operator.py
tests.contrib.hooks.test_gcp_container_hook.py

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/contrib/operators/gcp_container_operator.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+from airflow.utils.decorators import apply_defaults
+
+
+class GKEClusterDeleteOperator(BaseOperator):
+    template_fields = ['project_id', 'gcp_conn_id', 'location', 'api_version', 'body']
+
+    @apply_defaults
+    def __init__(self,
+                 project_id,
+                 cluster_name,
```
Comment (fenglu-db) [inline airflow/contrib/operators/gcp_container_operator.py]: s/cluster_name/name?

File: airflow/contrib/operators/gcp_container_operator.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    def _check_input(self):
+        if not self.project_id:
+            self.log.error('Please specify the project_id of cluster to use.')
+            raise AirflowException('project_id parameter missing, unable to trigger job')
+        if not self.cluster_name:
+            self.log.error('Please specify the cluster\'s name to delete.')
+            raise AirflowException('cluster_name parameter missing, unable to start job')
+        if not self.location:
+            self.log.error('Please specify the cluster location id to use.')
+            raise AirflowException('location parameter missing, unable to trigger job')
```
Comment (fenglu-db) [inline airflow/contrib/operators/gcp_container_operator.py]: nit: if not all([self.project_id, self.name, self.location]):
  self.log.error(...)
   raise AirflowException...

File: airflow/contrib/operators/gcp_container_operator.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    def execute(self, context):
+        self._check_input()
+        hook = GKEClusterHook(self.project_id, self.location)
+        delete_result = hook.delete_cluster(cluster_name=self.cluster_name)
```
Comment (fenglu-db) [inline airflow/contrib/operators/gcp_container_operator.py]: s/self.cluster_name/self.name

File: airflow/contrib/operators/gcp_container_operator.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        """
+        super(GKEClusterCreateOperator, self).__init__(*args, **kwargs)
+
+        self.project_id = project_id
+        self.gcp_conn_id = gcp_conn_id
+        self.location = location
+        self.api_version = api_version
+        self.body = body
+
+    def _check_input(self):
```
Comment (fenglu-db) [inline airflow/contrib/operators/gcp_container_operator.py:152]: _check_input can be pushed to the GkeClusterHook or an "internal" function in the operator module.

File: airflow/contrib/hooks/gcp_container_hook.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        :param timeout: The amount of time, in seconds, to wait for the request to
+        complete. Note that if retry is specified, the timeout applies to each individual
+        attempt.
+        :type timeout: float
+        :return: The full url to the newly created cluster if successful else None
+        """
+
+        if isinstance(cluster, dict):
+            cluster_proto = Cluster()
+            cluster = self._dict_to_proto(py_dict=cluster, proto=cluster_proto)
```
Comment (fenglu-db) [inline airflow/contrib/hooks/gcp_container_hook.py:170]: what if this call failed? would it throw an excpetion?

File: airflow/contrib/hooks/gcp_container_hook.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        try:
+            op = self.client.create_cluster(project_id=self.project_id,
+                                            zone=self.location,
+                                            cluster=cluster,
+                                            retry=retry,
+                                            timeout=timeout)
+            op = self.wait_for_operation(op)
+            return op.target_link
+        except AlreadyExists:
+            self.log.info(
```
Comment (fenglu-db) [inline airflow/contrib/hooks/gcp_container_hook.py]: please also return the cluster name similar to op.target_link.

**Comment by Noremac201** [comment]
Taken care of the issues pointed out, _check_input can't be pushed to a static method, or to the hook since what it checks is unique depending on operator. 
As for the exception, I've made it more apparent what happens in the docs, an exception is thrown if JSON parsing goes wrong.

Thanks

File: airflow/contrib/hooks/gcp_container_hook.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        worker nodes. Firewalls and routes that were configured during
+        cluster creation are also deleted. Other Google Compute Engine
+        resources that might be in use by the cluster (e.g. load balancer
+        resources) will not be deleted if they weren’t present at the
+        initial create time.
+
+        :param name: The name of the cluster to delete
+        :type name: str
+        :param retry: Retry object used to determine when/if to retry requests.
+        If None is specified, requests will not be retried.
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_container_hook.py]: can you add a tab here:
```
    If None is specified, requests will not be retried.
```

File: airflow/contrib/hooks/gcp_container_hook.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        :type retry: google.api_core.retry.Retry
+        :param timeout: The amount of time, in seconds, to wait for the request to
+        complete. Note that if retry is specified, the timeout applies to each
+        individual attempt.
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_container_hook.py]: Same.. tab here

File: airflow/contrib/hooks/gcp_container_hook.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        resources) will not be deleted if they weren’t present at the
+        initial create time.
+
+        :param name: The name of the cluster to delete
+        :type name: str
+        :param retry: Retry object used to determine when/if to retry requests.
+        If None is specified, requests will not be retried.
+        :type retry: google.api_core.retry.Retry
+        :param timeout: The amount of time, in seconds, to wait for the request to
+        complete. Note that if retry is specified, the timeout applies to each
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_container_hook.py]: Same.. tab here

File: airflow/contrib/hooks/gcp_container_hook.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+        Engine instances.
+
+        :param cluster: A Cluster protobuf or dict. If dict is provided, it must be of
+        the same form as the protobuf message google.cloud.container_v1.types.Cluster
+        :type cluster: dict or google.cloud.container_v1.types.Cluster
+        :param retry: A retry object (google.api_core.retry.Retry) used to retry requests.
+        If None is specified, requests will not be retried.
+        :type retry: google.api_core.retry.Retry
+        :param timeout: The amount of time, in seconds, to wait for the request to
+        complete. Note that if retry is specified, the timeout applies to each individual
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_container_hook.py]: Same.. tab here

File: airflow/contrib/operators/gcp_container_operator.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        :param name: The name of the resource to delete, in this case cluster name
+        :type name: str
+        :param location: The name of the Google Compute Engine zone in which the cluster
+        resides.
+        :type location: str
+        :param gcp_conn_id: The connection ID to use connecting to Google Cloud Platform.
+        :type gcp_conn_id: str
+        :param api_version: The api version to use
+        :type api_version: str
+        """
```
Comment (kaxil) [inline airflow/contrib/operators/gcp_container_operator.py:68]: Can you add an example usage as well?

**Review by kaxil — CHANGES_REQUESTED** [review]

**Comment by kaxil** [comment]
Thanks for this. This is certainly going to be very helpful.

**Comment by kaxil** [comment]
Can you please rebase against the master? The CI was failing for unrelated tests which are now fixed.

**Comment by Noremac201** [comment]
Rebased, looks like it's passing now.

Thanks!

File: airflow/contrib/operators/gcp_container_operator.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        **Operator Creation**: ::
+
+            operator = GKEClusterDeleteOperator(
+                        task_id='cluster_delete',
+                        project_id='my-project',
+                        location='cluster-location'
+                        name='cluster-name')
+
+        .. seealso::
+            For more detail on about creating clusters have a look at the reference:
```
Comment (kaxil) [inline airflow/contrib/operators/gcp_container_operator.py]: It is not creating right? This is deleting the cluster.

Comment (kaxil) [inline airflow/contrib/operators/gcp_container_operator.py]: It is not creating right? This is deleting the cluster.

**Review by kaxil — CHANGES_REQUESTED** [review]
And can you also please resolve the conflicts with `setup.py` after rebasing the master?

**Comment by kaxil** [comment]
@fenglu-g Any more comments on this?

**Comment by Noremac201** [comment]
Good catch -- fixed merge conflicts

Thanks

File: airflow/contrib/hooks/gcp_container_hook.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+            cluster.name))
+        try:
+            op = self.client.create_cluster(project_id=self.project_id,
+                                            zone=self.location,
+                                            cluster=cluster,
+                                            retry=retry,
+                                            timeout=timeout)
+            op = self.wait_for_operation(op)
+
+            return op.target_link
```
Comment (fenglu-db) [inline airflow/contrib/hooks/gcp_container_hook.py:189]: nit: op.target_link isn't the same as get_cluster output, good to be consistent.

File: airflow/contrib/hooks/gcp_container_hook.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+    def wait_for_operation(self, operation):
+        """
+        Given an operation, continuously fetches the status from Google Cloud until either
+        completion or an error occurring
+        :param operation: The Operation to wait for
+        :type operation: A google.cloud.container_V1.gapic.enums.Operator
+        :return: A new, updated operation fetched from Google Cloud
+        """
+        self.log.info("Waiting for OPERATION_NAME %s" % operation.name)
+        time.sleep(15)
```
Comment (fenglu-db) [inline airflow/contrib/hooks/gcp_container_hook.py]: nit: make 15 a named constant, something like OPERATION_POLL_INTERVAL

**Review by fenglu-db — COMMENTED** [review]
LGTM (some minor nits)

**Comment by fenglu-db** [comment]
@kaxil feel free to merge if it looks good to you as well, thank you for your help!

**Review by kaxil — APPROVED** [review]
LGTM as well. Thanks for adding this @Noremac201 and @fenglu-g for reviewing it.

**Pull request closed without merging**